### ISSUE TITLE
Trajectory implementations & literature

### ISSUE THREAD

**Issue by Daval-G** [mind-inria/mri-nufft#106]
The goal of this issue is to list and discuss trajectories from the literature. I gathered a bunch of articles, and I would be curious to know
- Which ones should be prioritized ? (would be nice for your future works)
- Which ones should not be reproduced ? (out-of-scope, too demanding computationally or in term of work, etc)

Here is a list of tasks related to trajectories from the literature that are not already implemented in MRI-NUFFT, along with a checklist for stuff that need further exploration:

## Trajectories
- 2D **TWIRL** from
    - "Twisting radial lines with application to robust magnetic resonance imaging of irregular flow", John I. Jackson et al. (1992)
- 3D **TPI** from
    - "Fast Three Dimensional Sodium Imaging", Fernando E. Boada (1997)
- <s>2D **Fibonacci spiral** from
    - "Uniform K-Space Sampling with an Interleaved Fibonacci Spiral Acquisition", Harvey E Cline and Thomas R Anthony (1999)</s>
- 2D **WHIRL** from
    - "An Optimized Center-Out k-Space Trajectory for Multishot MRI: Comparison With Spiral and Projection Reconstruction", James G. Pipe (1999)
- 2D **Teardrop** from
    - 'Teardrop, a novel non-raster readout for True FISP", Christopher Kumar Anad et al. (2001)
    - "Optimizing Teardrop, an MRI Sampling Trajectory", Christopher Kumar Anand et al. (2008)
- 3D **genetic** trajectory from
    - "Three dimensional k-space trajectory design using genetic algorithms", Sebastian Sabat et al. (2003)
- 3D **guided missile** trajectory from
    - "Fast Three Dimensional k-space Trajectory Design Using Missile Guidance", Ideas R. Mir et al. (2004)
- 3D radial trajectory from
    - "Fast isotropic volumetric coronary MR angiography using free‐breathing 3D radial balanced FFE acquisition", Stehning, C. et al.  (2004)
- Improve 2D waves based on **Bunched Phase Encoding** from
    - "Bunched Phase Encoding (BPE): A New Fast Data Acquisition Method in MRI", Hisamoto Moriguchi and Jeffrey L. Duerk (2006)
- 3D **genetic** trajectory from
    - "Random Volumetric MRI Trajectories via Genetic Algorithms", Andrew Thomas Curtis and Christopher Kumar Anand (2008)
- 3D **DURGA** from
    - "DURGA: A heuristically-optimized data collection strategy for volumetric Magnetic Resonance Imaging", Christopher Kumar Anand et al. (2008)
- 3D **Golden Means** from
    - "Temporal Stability of Adaptive 3D Radial MRI Using Multidimensional Golden Means", Rachel W. Chan (2009)
    - Golden Means from "Golden-Angle Radial MRI: Basics, Advances, and Applications", Li Feng (2022)
- 3D "Spiral" **Phyllotaxis** from
    - "Spiral Phyllotaxis: The Natural Way to Construct a 3D Radial Trajectory in MRI" Davide Piccini et al. (2011)
    - "Golden-Angle Radial MRI: Basics, Advances, and Applications", Li Feng (2022)
- 3D **RAZER** from
    - "RAZER: A Pulse Sequence for Whole-Brain Bolus Tracking at High Frame Rates", Sumeeth V. Jonathan et al. (2014)
- Improve 2D and 3D **concentric rings** based on
    - "Density-Weighted Concentric Circle Trajectories for High Resolution Brain Magnetic Resonance Spectroscopic Imaging at 7T", Lukas Hingerl et al. (2018)
    - "Three-dimensional, 2.5-minute, 7T phosphorus magnetic resonance spectroscopic imaging of the human heart using concentric rings", William T. Clarke (2022)
- 3D **T-Hex** spirals from
    - "T-Hex: Tilted hexagonal grids for rapid 3D imaging", Maria Engel et al. (2020)
    - "Mono-planar T-Hex: Speed and flexibility for high-resolution 3D imaging", Maria Engel et al. (2021)
- Reproduce **Looping Stars** from
    - "Looping Star fMRI in Cognitive Tasks and Resting State", Beatriz Dionisio-Parra et al. (2020)
- **AZTEK** tools and 3D trajectories from
    - "AZTEK: Adaptive zero TE k-space trajectories", Tanguy Boucneau et al. (2020)
- Distinguish 3D **Yarnball** from currently implemented 3D Seiffert spirals based on
    - "Quantifying Lung Water Density with Ultrashort Echo Time Yarn-Ball MRI", William Quinn Meadus (2020)
- 3D **REPI** from
    - "Three Dimensional Radial Echo Planar Imaging for Functional MRI", Christoph A. Rettenmeier et al. (2022)
- 2D **PILOT** from
    - "PILOT: Physics-Informed Learned Optimized Trajectories for Accelerated MRI", Tomer Weiss et al. (2021)
- 2D trajectory from
    - "Learning Optimal K-space Acquisition and Reconstruction using Physics-Informed Neural Networks", Wei Peng et al. (2022)
- 2D trajectory from
    - "Bayesian Optimization of Sampling Densities in MRI", Alban Gossard (2023)
- 2D **BJORK** from
    - "B-spline Parameterized Joint Optimization of Reconstruction and K-space Trajectories (BJORK) for Accelerated 2D MRI", Guanhua Wang et al. (2023)
- 3D **SNOPY** from
    - "Stochastic optimization of three-dimensional non-Cartesian sampling trajectory", Guanhua Wang et al. (2023)
- 2D and 3D **ECCENTRIC** from
    - "ECCENTRIC: a fast and unrestrained approach for high-resolution in vivo metabolic imaging at ultra-high field MR", Antoine Klauser et al. (2023)
- 3D **PETALUTE** from
    - "An Accelerated PETALUTE MRI Sequence for In Vivo Quantification of Sodium Content in Human Articular Cartilage at 3T", Cameron X. Villarreal et al. (2024)
    - "Accerated Preclinical UHF Abdominal T1 Mapping using Novel Rosette Accelerated Preclinical UHF Abdominal T1 Mapping using Novel Rosette Ultrashort Echo Time (PETALUTE) Ultrashort Echo Time (PETALUTE)", Alexandra Lipka et al. (2024)
- 3D **TURBINE** from
    - "Multi-shot 3D di usion MRI sequence for a fast and high-resolution imaging at 3T", Sajjad Feizollah et al. (2024)
- 3D **Arc-ZTE** from
    - "Arc-ZTE: Incoherent k-space sampling in time using continuously-slewed gradients for exible, dynamic, quiet Zero TE MRI", Shreya Ramachandran et al. (2024)

## Check-list
- "Optimal Design of k-Space Trajectories Using a MultiObjective Genetic Algorithm", Brian M. Dale et al. (2004)
- "Single TrAjectory Radial (STAR) Imaging", Gordon E. Sarty (2004)
- "A Fast Method for Designing Time-Optimal Gradient Waveforms for Arbitrary k-Space Trajectories", Michael Lustig et al. (2008)
- "Spiral Trajectory Design: A Flexible Numerical Algorithm and Base Analytical Equations" James G. Pipe et al. (2014)
- "Fast and Robust Design of Time-Optimal k-Space Trajectories in MRI", Mathias Davids et al. (2015)
- "Squeezed Trajectory Design for Peak RF and Integrated RF Power Reduction in Parallel Transmission MRI", Qing Li et al. (2018)
- "Partial fourier shells trajectory for non-cartesian MRI", Tao et al. (2019)
- "High resolution free-breathing respiratory-resolved volumetric lung imaging at 0.55T using stack-of-spiral out-in bSSFP", Ziwei Zhao et al. (2024)
- "Silent Looping Star fMRI with enhanced Encoding and Reconstruction Performance", Ana Beatriz Solana et al. (2024)
- "Fast and Motion-Robust Non-Contrast MR Angiography using Centric k –k Trajectory.", Vadim Malis et al. (2024)
- "ABRICOTINE MRI: Enhancing Sparsity Across the Three Dimensions of the Fourier Domain in Cartesian Sampling", Antoine Klauser et al. (2024)
- "3D UTE With Twisted Trajectories", Michael Carl et al. (2024)

**Comment by Daval-G** [comment]
For context, I already have advanced implementations of the following trajectories:
- 3D **genetic** trajectory from
    - "Three dimensional k-space trajectory design using genetic algorithms", Sebastian Sabat et al. (2003)
- 3D **guided missile** trajectory from
    - "Fast Three Dimensional k-space Trajectory Design Using Missile Guidance", Ideas R. Mir et al. (2004)

I also have works in progress for:
- 2D **TWIRL** from
    - "Twisting radial lines with application to robust magnetic resonance imaging of irregular flow", John I. Jackson et al. (1992)
- 2D **WHIRL** from
    - "An Optimized Center-Out k-Space Trajectory for Multishot MRI: Comparison With Spiral and Projection Reconstruction", James G. Pipe (1999)

Finally, I believe the following three trajectories are low-hanging fruits, if our focus is to reproduce literature:
- 3D radial trajectory from
    - "Fast isotropic volumetric coronary MR angiography using free‐breathing 3D radial balanced FFE acquisition", Stehning, C. et al. (2004)
- 3D **Golden Means** from
    - "Temporal Stability of Adaptive 3D Radial MRI Using Multidimensional Golden Means", Rachel W. Chan (2009)
Golden Means from "Golden-Angle Radial MRI: Basics, Advances, and Applications", Li Feng (2022)
- 3D "Spiral" **Phyllotaxis** from
    - "Spiral Phyllotaxis: The Natural Way to Construct a 3D Radial Trajectory in MRI" Davide Piccini et al. (2011)
"Golden-Angle Radial MRI: Basics, Advances, and Applications", Li Feng (2022)

**Comment by paquiteau** [comment]
Thanks for the detail review ! I edited you answer for linking to relevant issue and add check boxes.

**Comment by chaithyagr** [comment]
This is just amazing list! 
I think its best practice to have examples with respective papers within a folder called "examples/trajectories", to ensure we can have it visualized etc right within docs. WDYT?

**Comment by chaithyagr** [comment]
@paquiteau possibly some way to have nomenclature and trajectory repository? get_trajectory('name')(params)?

**Comment by paquiteau** [comment]
I don't see any difficulties for that. We can go two ways: 
- A decorator on top of each trajectory function using the `MethodRegister` helper (as for density compensation for instance)

- Use some import-time logic to register all trajectories methods. We have a consistent `initialize_<trajectory_name>` naming scheme, so this should be fairly easy. 

And we can even propose both! The advantage of the decorator is that it can be used in external packages to register extra trajectories.
